_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,244 @@
+
+# Deploying Cruise Control
+
+With the Kafka cluster running and the Zookeeper endpoint accessible, Cruise Control can now be started safely and begin sampling metrics and monitoring the cluster. To run on Kubernetes, Cruise Control must be formatted into a container image. One can do this using the following Dockerfile. Note that the Cruise Control configuration file, 'cruisecontrol.properties', must be altered to reference the appropriate Kafka bootstrap and Zookeeper service endpoints. In addition to pointing Cruise Control to the proper service endpoints, the Dockerfile sets up the Cruise Control GUI frontend to be available upon startup.
+
+```Dockerfile
+FROM centos:7
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
+
+RUN yum -y install git java-1.8.0-openjdk-devel && \
+    yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+WORKDIR cruise-control
+
+RUN ./gradlew jar copyDependantLibs
+
+# Configure Cruise Control to point to the proper Kafka and Zookeeper endpoints
+RUN sed -i 's/bootstrap.servers=.*/bootstrap.servers=my-cluster-kafka-bootstrap:9092/g' config/cruisecontrol.properties \
+    && sed -i 's/zookeeper.connect=.*/zookeeper.connect=zoo-entrance:2181/g' config/cruisecontrol.properties \
+    && sed -i 's/capacity.config.file=.*/capacity.config.file=config\/capacityJBOD.json/g' config/cruisecontrol.properties \
+    && sed -i 's/sample.store.topic.replication.factor=.*/sample.store.topic.replication.factor=1/g' config/cruisecontrol.properties 
+
+# Ensure Cruise Control log directory is writable
+RUN mkdir logs \ 
+    && touch ./logs/kafkacruisecontrol.log \
+    && chmod a+rw -R .
+
+# Install Cruise Control GUI Frontend
+RUN curl -L https://github.com/linkedin/cruise-control-ui/releases/download/v0.1.0/cruise-control-ui.tar.gz \
+    -o /tmp/cruise-control-ui.tar.gz \
+    && tar zxvf /tmp/cruise-control-ui.tar.gz
+
+ENTRYPOINT ["/bin/bash", "-c", "./kafka-cruise-control-start.sh config/cruisecontrol.properties"]
+```
+Build the image:
+```
+docker build . -t <cruise-control-image-name>
+```
+Then, using the following file, 'cruise-control-deployment.yaml', create a Cruise Control deployment that references the Cruise Control image, 'cruise-control-image-name', and a Cruise Control service endpoint for enabling communication with Cruise Control from outside the Cruise Control pod.
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: my-cruise-control
+  labels:
+    app: cruise-control
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: cruise-control
+        name: my-cruise-control
+    spec:
+      containers:
+      - name: my-cruise-control
+        image: <cruise-control-image-name>
```

Comment:
This will nto work for most people with Minikube etc. We should at least mention that they will need to push it somewhere etc.